### PR TITLE
TypeScript project references

### PR DESCRIPTION
**Pull request by milesj**
Automates TS project references for workspace enabled projects via the new `--reference-workspaces` option.

### CODE REVIEW

File: packages/driver-typescript/src/TypeScriptDriver.ts [hunk 1]
```diff
@@ -41,4 +56,138 @@ export default class TypeScriptDriver extends Driver<TypeScriptConfig> {
...
+    );
+
+    delete config.compilerOptions;
+
+    // Generate references and update paths
+    config.extends = './tsconfig.options.json';
+    config.files = [];
+    config.references = this.tool.getWorkspacePackages({ root: workspaceRoot }).map(wsPkg => ({
+      path: path.relative(workspaceRoot, wsPkg.workspace.packagePath),
+    }));
```
Comment (hayes) [inline packages/driver-typescript/src/TypeScriptDriver.ts]:
```ts
    config.references = []
        
    const { testsFolder } = this.options;
    this.tool.getWorkspacePackages({ root: workspaceRoot }).forEach(wsPkg => {
      config.references.push({ path: path.relative(workspaceRoot, wsPkg.workspace.packagePath) })
      
      if (testsFolder) {
         const testsPath = path.join(wsPkg.workspace.packagePath, testsFolder);
         
         if (fs.existsSync(testsPath) {
           config.references.push({ path: path.relative(workspaceRoot, testsPath) })
         }
    })
```

Comment (milesj) [reply packages/driver-typescript/src/TypeScriptDriver.ts]: This works unless a package declares custom type declaration files, which are then outside of the tests "root". >_<

Like here: https://github.com/beemojs/beemo/tree/master/packages/core/types/hygen

Trying to figure this out now.

Comment (milesj) [reply packages/driver-typescript/src/TypeScriptDriver.ts]: I also seem to hit this a lot... Which is unfortunate.

```
/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25865
                ts.Debug.assert(source !== target);
                         ^
Error: Debug Failure. False expression.
    at mergeSymbol (/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25865:26)
    at /Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25964:47
    at Map.forEach (<anonymous>)
    at mergeSymbolTable (/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25962:20)
    at mergeSymbol (/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25887:21)
    at /Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25964:47
    at Map.forEach (<anonymous>)
    at mergeSymbolTable (/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25962:20)
    at mergeSymbol (/Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25892:21)
    at /Users/miles_johnson/Sites/beemo/node_modules/typescript/lib/tsc.js:25964:47
```

Comment (milesj) [reply packages/driver-typescript/src/TypeScriptDriver.ts]: Actually, it was the config upstream in `@milesj` causing it. However, builds are much slower including tests also.

Comment (hayes) [reply packages/driver-typescript/src/TypeScriptDriver.ts]: thats lame...

**Pull request merged**